rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
+  "args": {
+    "assert": [
+      // refers to the invocation from the example
+      { "/": "bafy..allocate" },
+      // refers to the receipt corresponding to the above invocation
+      { 
+        "out": {
+          "ok": {
+            // result of the add is the content (location) commitment
+            // that is produced as result of "bafy..accept"
+            "size": 2_097_152,
+            "address": {
+              "url": "https://node1.storage/upload/xxxa-presigned",    
+              "headers": {},
+              "expires": 1712903125
+            }
+          }
+        },
+        "next": []
+      },
+    ],
+  },
+}
+```
+
+Legacy receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "ran": {
+    "/": "bafyreia5tctxekbm5bmuf6tsvragyvjdiiceg5q6wghfjiqczcuevmdqcu"
+  },
+  "out": {
+    "ok": {
+      "size": 2_097_152,
+      "address": {
+        "url": "https://node1.storage/upload/xxxa-presigned",    
+        "headers": {},
+        "expires": 1712903125
+      }
+    }
+  },
+  "fx": [],
+}
+```
+
+4. On success, router creates a success receipt for the original user
+
+Receipt:
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:web3.storage",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
+  "args": {
+    "assert": [
+      // refers to the invocation from the example
+      { "/": "bafy..allocate" },
+      // refers to the receipt corresponding to the above invocation
+      { 
+        "out": {
+          "ok": {
+            "size": 2_097_152,
+            "address": {
+              "url": "https://node1.storage/upload/xxxa-presigned",    
+              "headers": {},
+              "expires": 1712903125
+            }
+          }
+        },
+        "next": []
+      },
+    ],
+  },
+}
+```
+
+Legacy receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:web3.storage",
+  "ran": {
+    "/": "bafy..allocate"
+  },
+  "out": {
+    "ok": {
+      // result of the add is the content (location) commitment
+      // that is produced as result of "bafy..accept"
+      "size": 2_097_152,
+      "address": {
+        "url": "https://node1.storage/upload/xxxa-presigned",    
+        "headers": {},
+        "expires": 1712903125
+      }
+    }
+  },
+  "fx": [],
+}
+```
+
+The router should also store a BlobAllocate record:
+
+```json
+{
+  "did": "did:web:node1.storage",
+  "blob": {
+    // multihash of the blob as byte array
+    "digest": { "/": { "bytes": "mEi...sfKg" } },
+    // size of the blob in bytes
+    "size": 2_097_152,
+  },
+}
+```
+
+4. On fail, router adds the failed nodes to a list of "alreadyFailed", and returns to step 2.
+
+5. If all nodes fail, return a fail receipt to the user for `blob/allocate`
```

Comment:
I think what I'm failing to understand if we expect to get `n` allocations or if we just trying to get a single allocation.

Reply:
this is a loop until success operation.

Reply:
Sounds like a single success, correct ?